lib/Dialect/EmitC/Conversion/TosaToEmitC.cpp

```diff
@@ -108,11 +108,169 @@ class RsqrtOpConversion : public OpConversionPattern<tosa::RsqrtOp> {
+    StringAttr callee = rewriter.getStringAttr(funcName);
+    ArrayAttr args;
+
+    SmallVector<Attribute, 4> templateArgs_;
+
+    if (explicitResultType) {
+      Type type = srcOp.getType();
+      templateArgs_.push_back(TypeAttr::get(type));
+    }
+
+    if (explicitOperandTypes) {
+      for (auto &operand : operands) {
+        Type type = operand.getType();
+        templateArgs_.push_back(TypeAttr::get(type));
+      }
+    }
+
+    ArrayAttr templateArgs = ArrayAttr::get(srcOp.getContext(), templateArgs_);
+
+    SmallVector<Value, 2> broadcastedOperands =
+        create_broadcast_op_if_needed(srcOp, operands, rewriter);
+
+    rewriter.replaceOpWithNewOp<emitc::CallOp>(
+        srcOp, srcOp.getType(), callee, args, templateArgs,
+        mlir::ValueRange({broadcastedOperands[0], broadcastedOperands[1]}));
+
+    return success();
+  }
+
+  StringRef funcName;
+  // If set, use the result type of the operation as template parameter
+  bool explicitResultType;
+  // If set, use the operand types as (additional) template parameters
+  bool explicitOperandTypes;
+};
+
+class MulOpConversion : public OpConversionPattern<tosa::MulOp> {
+  using OpConversionPattern<tosa::MulOp>::OpConversionPattern;
+
+public:
+  MulOpConversion(MLIRContext *ctx, StringRef funcName,
+                  bool explicitResultType = false,
+                  bool explicitOperandTypes = false)
+      : OpConversionPattern<tosa::MulOp>(ctx), funcName(funcName),
+        explicitResultType(explicitResultType),
+        explicitOperandTypes(explicitOperandTypes) {}
+
+private:
+  LogicalResult
+  matchAndRewrite(tosa::MulOp mulOp, ArrayRef<Value> operands,
+                  ConversionPatternRewriter &rewriter) const override {
+    StringAttr callee = rewriter.getStringAttr(funcName);
+
+    SmallVector<Attribute, 4> templateArgs_;
+
+    if (explicitResultType) {
+      Type type = mulOp.getType();
+      templateArgs_.push_back(TypeAttr::get(type));
+    }
+
+    if (explicitOperandTypes) {
+      for (auto &operand : operands) {
+        Type type = operand.getType();
+        templateArgs_.push_back(TypeAttr::get(type));
+      }
+    }
+
+    SmallVector<Attribute, 1> args_;
+    args_.push_back(mulOp.shiftAttr());
+    ArrayAttr args = rewriter.getArrayAttr(args_);
+
+    ArrayAttr templateArgs = ArrayAttr::get(mulOp.getContext(), templateArgs_);
+
+    SmallVector<Value, 2> broadcastedOperands =
+        create_broadcast_op_if_needed(mulOp, operands, rewriter);
+
+    rewriter.replaceOpWithNewOp<emitc::CallOp>(
+        mulOp, mulOp.getType(), callee, args, templateArgs,
+        mlir::ValueRange({broadcastedOperands[0], broadcastedOperands[1]}));
+
+    return success();
+  }
+
+  StringRef funcName;
+  // If set, use the result type of the operation as template parameter
+  bool explicitResultType;
+  // If set, use the operand types as (additional) template parameters
+  bool explicitOperandTypes;
```

Comment:
As this conversion operates on a single op, these attributes can be removed (inlined).